functionary/ui/views/tasks.py

```diff
@@ -1,10 +1,138 @@
+    except Exception:
+        return False
+
+    return True
+
+
+def _format_csv_table(result: str) -> dict:
+    """Convert a string of CSV into a format suitable for table rendering"""
+    csv_data = csv.reader(result.splitlines())
+    formatted_result = {
+        "headers": next(csv_data),
+        "data": [row for row in csv_data],
+    }
+
+    return formatted_result
+
+
+def _format_json_table(result: list) -> dict:
+    """Convert a JSON list into a format suitable for table rendering"""
+    headers = [key for key in result[0].keys()]
+    res_data = []
+
+    for row in result:
+        r = []
+        for key in headers:
+            r.append(row[key])
+        res_data.append(r)
+
+    return {"headers": headers, "data": res_data}
+
+
+def _format_table(result):
+    """Convert a result to a table friendly format
+
+    This will take in a "string" or "json" result and return
+    the headers and a list of rows for the data. A result_type of
+    string is assumed to be csv formatted data - a csv reader
+    parses the data into the resulting header and data.
+
+    A result type of json should be of the following format:
+    [
+        {"prop1":"value1", "prop2":"value2"},
+        {"prop1":"value3", "prop2":"value4"}
+    ]
+    The headers are derived from the keys of the first entry in the list.
+    """
+    if type(result) is str:
+        return _format_csv_table(result)
+    elif type(result) is list and len(result) > 0:
+        return _format_json_table(result)
+    else:
+        raise ValueError("Unable to convert result to table")
+
+
+def _show_output_selector(result, result_type, completed):
+    """Determines if the output format selector should be rendered"""
+    if not completed:
+        return False
+
+    if result_type is list:
+        show_selector = True
+    elif result_type is str:
+        show_selector = _detect_csv(result)
+    else:
+        show_selector = False
+
+    return show_selector
+
+
+def _format_result(result, result_type, format):
+    """Inspects the task result and formats the result data for in the desired format
+    as appropriate"""
+    output_format = "table"
+    formatted_result = None
+    format_error = None
+
+    match format:
+        case "display_raw":
+            if result_type in [list, dict]:
+                output_format = "json"
+            else:
+                output_format = "string"
+        case "display_table":
+            try:
+                formatted_result = _format_table(result)
+            except Exception:
+                format_error = "Result data is unsuitable for table output"
+        case _:
+            raise BadRequest("Invalid display format")
+
+    return output_format, format_error, formatted_result
+
+
+def _get_result_context(context, format):
+    task = context["task"]
+
+    completed = task.status in FINISHED_STATUS
+    result_type = type(task.result)
+
+    output_format, format_error, formatted_result = _format_result(
```

Comment:
I didn't like doing the type check twice since they're internal functions called from one space. I moved the check into the functions and only do it in _format_results when it's actually needed.